Adding tenant to stores and APIs. To do this, introducing the

albertshau opened the pull request:
concept of an Account to keep track of all user information, and
views for the cluster and entity stores. Views are per account,
and only allow access to parts of the database that can be
accessed by that account.

albertshau committed:
Adding tenant to stores and APIs. To do this, introducing the
concept of an Account to keep track of all user information, and
views for the cluster and entity stores. Views are per account,
and only allow access to parts of the database that can be
accessed by that account.

poornachandra commented on server/src/main/java/com/continuuity/loom/account/Account.java:
```diff
@@ -0,0 +1,80 @@
+  private final String userId;
+  private final String tenantId;
+  @Expose(serialize = false, deserialize = false)
+  private final boolean isSystem;
```
Can you add a comment on what isSystem means, and why it should not be serialized?

albertshau replied:
done

poornachandra commented on server/src/main/java/com/continuuity/loom/store/cluster/SQLClusterStoreService.java:
```diff
@@ -0,0 +1,521 @@
+      DBQueryHelper.createDerbyTable("CREATE TABLE nodes ( cluster_id BIGINT, id VARCHAR(64), node BLOB )",
+                                     dbConnectionPool);
+    }
+    systemView = new SQLSystemClusterStoreView(dbConnectionPool, Account.SYSTEM_ACCOUNT);
```
This can probably move to constructor.

albertshau replied:
done

poornachandra commented on server/src/main/java/com/continuuity/loom/store/cluster/SQLClusterStoreService.java:
```diff
@@ -0,0 +1,521 @@
+    try {
+      Connection conn = dbConnectionPool.getConnection();
+      try {
+        PreparedStatement statement = conn.prepareStatement("SELECT job FROM jobs WHERE job_num=? AND cluster_id=?");
```
Can you add a comment why job id is not filtered by tenant?

albertshau replied:
as mentioned further down, will move these methods into a separate interface that only the system will use.

poornachandra commented on server/src/main/java/com/continuuity/loom/store/entity/SQLEntityStoreService.java:
```diff
@@ -0,0 +1,67 @@
+import java.sql.Statement;
+
+/**
+ *
```
Empty javadoc

albertshau replied:
done

poornachandra commented on server/src/main/java/com/continuuity/loom/store/cluster/BaseSQLClusterStoreView.java:
```diff
@@ -0,0 +1,222 @@
+  abstract PreparedStatement getSetClusterStatement(
+    Connection conn, long id, Cluster cluster, ByteArrayInputStream clusterBytes) throws SQLException;
+
+  abstract PreparedStatement getInsertClusterStatement(
```
getInsertClusterStatement can be implemented in the base class since it has the same implementation across all derived classes, right?

albertshau replied:
true, will move it

poornachandra commented on server/src/main/java/com/continuuity/loom/store/cluster/ClusterStoreService.java:
```diff
@@ -0,0 +1,118 @@
+import java.util.Set;
+
+/**
+ *
```
Empty javadoc

albertshau replied:
done

poornachandra commented on server/src/main/java/com/continuuity/loom/store/cluster/ClusterStoreService.java:
```diff
@@ -0,0 +1,118 @@
+   * @return The cluster job with the given id, or null if none exists.
+   * @throws IOException if there was a problem getting the cluster job.
+   */
+  ClusterJob getClusterJob(JobId jobId) throws IOException;
```
Maybe move all methods that don't need a view to a different interface so that this interface is consistent with EntityStoreService? Also, the methods are used by different entities.

albertshau replied:
agreed, these methods are never called by a user and would be cleaner to have them separate

poornachandra commented on server/src/main/java/com/continuuity/loom/store/cluster/BaseSQLClusterStoreView.java:
```diff
@@ -0,0 +1,222 @@
+
+  @Override
+  public void deleteCluster(String clusterId) throws IOException {
+    long clusterNum = Long.parseLong(clusterId);
```
writeCluster throws an exception when a user is not allowed to write to a cluster, but deleteCluster silently ignores it. We need to make the behavior uniform.

albertshau replied:
the difference is that write has a cluster object with an associated account when it is doing the writing.  Delete is just given an id, and I don't think we should distinguish between a user trying to delete a nonexistant cluster versus one they don't have access to.

poornachandra commented on server/src/main/java/com/continuuity/loom/store/cluster/BaseSQLClusterStoreView.java:
```diff
@@ -0,0 +1,222 @@
+ */
+public abstract class BaseSQLClusterStoreView implements ClusterStoreView {
+  private final DBConnectionPool dbConnectionPool;
+  private static final JsonSerde codec = new JsonSerde();
```
static final should be all CAPS.

albertshau replied:
done

poornachandra commented:
Mostly reviewed the new classes added, since others were just refactoring. Minor comments, please fix and merge.

albertshau committed:
pull request feedback.  Moving node, job, task calls out of
the ClusterStoreService and putting it in a ClusterStore class
that is able to see the full store.

albertshau committed:
remove unused imports.
